src/Microsoft.FeatureManagement/FeatureManagerSnapshot.cs

```diff
@@ -13,8 +14,8 @@ namespace Microsoft.FeatureManagement
     class FeatureManagerSnapshot : IFeatureManagerSnapshot
     {
         private readonly IFeatureManager _featureManager;
-        private readonly IDictionary<string, bool> _flagCache = new Dictionary<string, bool>();
-        private IEnumerable<string> _featureNames;
+        private readonly ConcurrentDictionary<string, Task<bool>> _flagCache = new ConcurrentDictionary<string, Task<bool>>();
```

Comment:
FWIW this isn't actually race proof. `ConcurrentDictionary.GetOrAdd` has atomic publish semantics; not atomic execution semantics. If you want to prevent duplicate executions for costly feature flag evaluations, you have to cache a `Lazy<>` with its mode set to `ExecutionAndPublication` instead.

Basically:
```cs
public Task<bool> IsEnabledAsync(string feature)
{
  return _flagCache
    .GetOrAdd(feature, (key, arg) => new Lazy<Task<bool>>(
      async () => await arg.IsEnabledAsync(key).ConfigureAwait(false),
      LazyThreadSafetyMode.ExecutionAndPublication
    ), _featureManager )
    .Value;
}
```

Reply:
Thanks @rjgotten for the suggested improvement.